### DISCUSSION TITLE
When might I want to use removeQueries instead of invalidateQueries and vice-versa

### DISCUSSION THREAD

**Discussion by Aaronius** [TanStack/query#3169] [Q&A] [answered]
The [`removeQueries` documentation](https://react-query.tanstack.com/reference/QueryClient#queryclientremovequeries) is a bit limited. It isn't really clear what effect calling `removeQueries` might have on active or inactive queries (maybe none?) or why I might want to use `invalidateQueries` instead. Could someone give me a brief primer please?

**Comment by TkDodo** [comment, marked as answer]
`removeQueries` has no effect on active queries. If a component re-renders for some reason that then tries to use the removed query, it will go into hard loading state again. Not sure why you'd ever want to remove an active query...

`invalidateQueries` is something completely different. It:
- marks all matching queries as `stale`, so that they will be refetched the next time they are used
- active queries will be immediately refetched per default (unless you change the options)

invalidation _never_ removes something from the cache, it just refetches.

**Reply by Aaronius** [reply]
Thanks @TkDodo. I think I'm not really understanding how a query in cache marked as stale affects components using a `useQuery` hook. I see you mentioned _so that they will be refetched the next time they are used_ but I'm maybe I'm not understanding what _the next time they are used_ means. I don't think I exactly understand what "active query" means either.

To illustrate, we have an app with two routes (using react-router). Assuming the user goes to Route A, we render a component that uses `useQuery` with a query key of `todos`. The data loads and it's rendered. The user then goes to Route B and clicks a button. If, in the button's `onClick` handler, we call `removeQueries`, and the user navigates back to Route A, they don't see any stale data. The query is immediately in the loading state. However, if we instead call `invalidateQueries` from the button's `onClick` handler and the user navigates back to Route A, they see stale data for a brief moment before fresh content is fetched. We'd prefer the user never sees stale data in this case, so we're leaning toward calling `removeQueries`, but I think we might just be misunderstanding `invalidateQueries`. Is this behavior we're seeing expected? Perhaps something else in our app code is playing into this; I haven't tried creating a super simple reproduction.

**Reply by TkDodo** [reply]
> I see you mentioned so that they will be refetched the next time they are used but I'm maybe I'm not understanding what the next time they are used means.

the points when stale queries will be refetched are described here: https://react-query.tanstack.com/guides/important-defaults
there is also a good example in the docs: https://react-query.tanstack.com/guides/caching

lastly, I've blogged a lot about this topic already, I can suggest starting here: https://tkdodo.eu/blog/practical-react-query

**Reply by Aaronius** [reply]
Thanks @TkDodo!

**Reply by gulshairbutt** [reply]
Thanks @Aaronius for the response. I looked into the referenced links, I am still not sure about where to use the `invalidateQueries` and `removeQueries`. 

e.g. https://tanstack.com/query/latest/docs/framework/react/guides/query-invalidation here it tells me how to use them but not why and where to use them.


But through my experimentation, I came to the following conclusions:
1. `invalidateQueries` marks the data stale and properly refetches the data if the query is currently active.
2. `removeQueries` removes the cache and properly refetches the data when the query is active again.

But if I try the `invalidateQueries` for inactive queries, it shows me the stale data even if I set the `gcTime` to 0. 

Now, my question is that is this the right way to use it or am I missing something?

Thank you in advance..!!

**Reply by TkDodo** [reply]
> I am still not sure about where to use the invalidateQueries and removeQueries.

You probably never need `removeQueries`, unless when the user logs out and you want to clear the whole cache or all queries that contain data belonging to that user.

invalidation doesn’t remove anything so the two aren’t even remotely comparable in what they do.

> But if I try the invalidateQueries for inactive queries, it shows me the stale data

if you invalidate an inactive query, it will be marked as invalidated and will be refetched the next time it becomes used. Invalidation doesn’t remove anything so until new data comes in, you will see stale data.

> even if I set the gcTime to 0.

if you have a `gcTime` of `0` and the query becomes inactive, it will be immediately removed from the cache. After that, you can no longer target it with `invalidateQueries` because it doesn’t exist in the cache. You can’t invalidate something that isn’t there.

**Reply by gulshairbutt** [reply]
Hmm, that's the problem. It doesn't refetches when the query is active again. I used the dev tools and was able to see that the query data is stale (via invalidation), and the query is active again but no new request triggers to fetch the new data.

**Reply by TkDodo** [reply]
> It doesn't refetches when the query is active again

Then you probably have changed some other setting, like `refetchOnMount`, to `false`. But I will not continue guessing. Please show a minimal, running, standalone reproduction.

**Reply by giacomocerquone** [reply]
in case of non-idempotent requests and central redirection logics based on the response of the query, that needs to be invalidated after a mutation and refetched with a different key, I can assure you removing active queries is not only useful, but leads to crazy clean code.

Thanks for thinking this out and for making such a thorough and complete library for data-fetching.

If you want me to delve on the situation I'm in just ask and I'll provide a longer explanation

PS:

yes the system may be ill designed and the react code fragile (mainly because after removing if you have a rerender it will still refetch the data) but still what I appreciate about libraries like yours is how thoroughly the devs needs are taken care of by preventing big changes to possibly legacy systems just because "library x wanted you to make this thing in this way only and if it's not ok, fork the code"

**Reply by gulshairbutt** [reply]
@TkDodo you are totally right. I was getting global prop `refetchOnMount=false`. But I would still remove it because it shows me cached data until new request is triggered. 

I really appreciate your time to answer the questions.

**Reply by TkDodo** [reply]
@giacomocerquone please do elaborate because Queries should likely always be idempotent ...

**Comment by nathanhannig** [comment]
If you edit some data related to a record, and return the user back to the record's page, removeQueries() works well so you don't show them the previous version of data they just edited while the query is refetching.

**Reply by TkDodo** [reply]
if you want that, I would use `resetQueries` because it does remove + re-render observers (and thus refetch) at the same time.

**Reply by vincerubinetti** [reply]
Took me a long time to figure out `reset` is what I needed. Boy, maybe I'm dumb but RQ is really confusing with the remove vs reset vs invalidate vs stale vs refetch vs etc. So many states and subtle distinctions to learn about.

**Reply by Aaronius** [reply]
The subtle nuances have bitten us numerous times as well.

**Reply by karasavm** [reply]
@TkDodo An article explaining remove vs reset vs invalidate vs stale vs refetch vs etc. would be a very good idea. There are many cases when mutate Object1 and onSuccess I setQeueryData to undefined on Object2 (related to Object1) and then invalidateQuery([Object1]). Because Object1 data are wrong after Object2 mutation. Is there a simpler way to do this.?

**Reply by TkDodo** [reply]
> and onSuccess I setQeueryData to undefined

tha's a noop in v4 because undefined is not valid for successful queries. I'm having a hard time imagining what you're trying to do so can't really help. State the problem and then we can try to find a solution

I've never needed remove or reset, barely ever touched refetch, and mainly use `invalidateQueries` only 🤷

**Reply by vincerubinetti** [reply]
I don't understand how you could never need "remove"... What if you do a mutation that modifies a record on the server, and you want to make RQ freshly refetch that record the next time it is accessed... isn't "remove" what clears that entry from the RQ cache? It certainly seems that way to me, as none of the other variants worked for me.

**Reply by vincerubinetti** [reply]
> invalidation never removes something from the cache, it just refetches.

This is where the terminology RQ uses becomes really confusing to me. "Refetch" to me means making another network request to get a new/fresh response and data, but here you're saying it refetches but doesn't clear the cache. So to me it sounds like you're saying, RQ makes a new network request, but then ignores it and still uses a cached response anyway, which doesn't make sense.

Maybe some kind of diagram of how all these terminologies and subtle distinctions fit together could be helpful. I think the RQ documentation assumes a lot of this is intuitive, ubiquitous terminology and concepts, but I'm not sure that it is.

**Reply by TkDodo** [reply]
> What if you do a mutation that modifies a record on the server, and you want to make RQ freshly refetch that record the next time it is accessed

you're describing exactly what `invalidateQueries` does.

> isn't "remove" what clears that entry from the RQ cache?

yes, it removes it, which means the next time you use it, you get a "hard loading state". With invalidation, you'll still see the stale data, but because it is marked as `stale`, it will be refetched in the background and you'll see new data.

There's a reason why the docs and all examples focus on this pattern rather than doing things like remove / reset / refetch: It has the best UX and is the easiest to get right for developers.

> "Refetch" to me means making another network request to get a new/fresh response and data, but here you're saying it refetches but doesn't clear the cache

yes, refetch does make another network request. But why do you need to _clear_ the cache before you make that network request? Maybe that's where the confusion comes from?

> So to me it sounds like you're saying, RQ makes a new network request, but then ignores it and still uses a cached response anyway, which doesn't make sense.

No, that's not what I'm saying. What it does is it:
- gives you the old, cached data
- makes a network request
- then gives you the new data

As opposed to:
- gives you nothing so you have to render a loading spinner
- makes a network request
- then gives you the new data

This is _literally_ what happens when you click the `refetch` button in the devtools. We don't delete data. We just replace it with new data once it comes in.

**Reply by graup** [reply]
> With invalidation, you'll still see the stale data, but because it is marked as stale, it will be refetched in the background and you'll see new data.

This means that until it's refetched, you see stale data. Sometimes that's fine but sometimes you want to make sure users don't see any stale data. Then you have to `remove`, right? Since a query object only has `remove` and `refetch` (and nothing like `reset`), does it make sense to call both of them? If my mutation is in the same component as the query, is it ok to `remove()` and `refetch()` in the mutation's onSuccess?

**Reply by TkDodo** [reply]
use queryClient.resetQueries() if you want that

**Reply by jrnxf** [reply]
@TkDodo would love your take on this:

I think the new hydration changes to react query v5 expose a valid use case for remove queries IMO. Since hydration changes now occur in an effect, if I redirect a user to a route thats entry point is an RSC that prefetches data, and I know that query might be cached and I want to avoid a flicker of stale data (bc of hydration in a useEffect) removing the potentially cached query before navigating knowing the RSC will prefetch and supply seems like an elegant solution. 

Do you disagree?

**Reply by jrnxf** [reply]
> you're describing exactly what `invalidateQueries` does.

Not exactly - invalidateQueries will clear cache and immediately refetch, but you might not need to refetch right then and there, at which point removeQueries would be preferable. Niche use case maybe, but worth mentioning

**Reply by therour** [reply]
> invalidateQueries will clear cache and immediately refetch

@jrnxf it is immediately refetch because you have an active (or observed) query in the current rendered page.

**Reply by gulshairbutt** [reply]
I agree, I did quite a lot of experimentation around it. `invalidateQueries` work properly if a query is currently active, however if we have some edit form and this data is already cached on some other page (the query isn't currently active) it shows me the stale data even if I invalidated the data using `invalidateQueries`.  But `removeQueries` works for such cases perfectly.

If I am not wrong, the documentation isn't that clear about it. Or am I misunderstanding something?
